File: sdk/featureflags_sdk.py
"""
Python SDK for the Feature Flag & Traffic Control Platform.

Install dependencies:
    pip install requests sseclient-py

Usage:
    from featureflags_sdk import FeatureFlagsClient

    client = FeatureFlagsClient(base_url="http://localhost:8080", env="production")

    # One-shot fetch
    flags = client.get_all_flags()

    # Background polling
    client.poll_flags(interval=10)

    # Real-time SSE hot-reload
    client.start_sse(on_update=lambda flags: print("Updated:", flags))

    # Evaluate a flag for a user
    enabled = client.is_enabled("dark-mode", user_id="alice", tenant_id="acme")

    # Get an A/B variant
    variant = client.get_variant("button-color", user_id="alice")

    # Check rate limit
    allowed = client.check_rate_limit(route="/api/search", user_id="alice")
"""

import json
import importlib
import threading
import time

import requests


class FeatureFlagsClient:
    def __init__(self, base_url: str, env: str = "production", auth_token: str = ""):
        self.base_url = base_url.rstrip("/")
        self.env = env
        self.auth_token = auth_token
        self.flags: list = []
        self._lock = threading.Lock()

    def _auth_headers(self) -> dict:
        if not self.auth_token:
            return {}
        return {"Authorization": f"Bearer {self.auth_token}"}

    # ------------------------------------------------------------------
    # Flag listing
    # ------------------------------------------------------------------

    def get_all_flags(self) -> list:
        """Fetch all flags for the configured environment."""
        resp = requests.get(f"{self.base_url}/flags/all", params={"env": self.env}, timeout=5)
        resp.raise_for_status()
        with self._lock:
            self.flags = resp.json()
        return self.flags

    def poll_flags(self, interval: int = 10) -> None:
        """Start a background thread that refreshes flags every `interval` seconds."""
        def _loop():
            while True:
                try:
                    self.get_all_flags()
                except Exception as exc:
                    print(f"[FeatureFlagsClient] poll error: {exc}")
                time.sleep(interval)

        t = threading.Thread(target=_loop, daemon=True)
        t.start()

    # ------------------------------------------------------------------
    # SSE hot-reload
    # ------------------------------------------------------------------

    def start_sse(self, on_update=None) -> None:
        """
        Open an SSE connection and call `on_update(flags)` whenever a flag
        change is pushed from the control plane. Reconnects automatically.
        Runs in a background daemon thread.
        """
        try:
            sseclient = importlib.import_module("sseclient")  # pip install sseclient-py
        except ImportError as e:
            raise ImportError("pip install sseclient-py") from e

        def _listen():
            while True:
                try:
                    url = f"{self.base_url}/flags/stream"
                    resp = requests.get(url, params={"env": self.env}, stream=True, timeout=None)
                    client = sseclient.SSEClient(resp)
                    for event in client.events():
                        if event.event == "update":
                            try:
                                data = json.loads(event.data)
                            except json.JSONDecodeError:
                                data = None
                            if not isinstance(data, list):
                                data = self.get_all_flags()
                            with self._lock:
                                self.flags = data
                            if on_update:
                                on_update(self.flags)
                        # "ping" events are keepalives — ignored
                except Exception as exc:
                    print(f"[FeatureFlagsClient] SSE error: {exc}; reconnecting in 5s")
                    time.sleep(5)

        t = threading.Thread(target=_listen, daemon=True)
        t.start()

    # ------------------------------------------------------------------
    # Flag evaluation
    # ------------------------------------------------------------------

    def is_enabled(self, flag_name: str, user_id: str = "", tenant_id: str = "",
                   headers: dict = None) -> bool:
        """Evaluate a feature flag for a specific user context."""
        payload = {"UserID": user_id, "TenantID": tenant_id, "Headers": headers or {}}
        resp = requests.post(
            f"{self.base_url}/flags/{flag_name}/evaluate",
            json=payload,
            timeout=5,
        )
        if not resp.ok:
            return False
        return resp.json().get("enabled", False)

    # ------------------------------------------------------------------
    # A/B experiments
    # ------------------------------------------------------------------

    def get_variant(self, experiment_name: str, user_id: str) -> str:
        """Get the deterministic variant assigned to a user for an experiment."""
        resp = requests.get(
            f"{self.base_url}/experiment/{experiment_name}/variant",
            params={"userId": user_id},
            timeout=5,
        )
        resp.raise_for_status()
        return resp.json().get("variant", "")

    def record_conversion(self, experiment_name: str, variant: str) -> None:
        """Record a conversion event for an experiment variant."""
        requests.post(
            f"{self.base_url}/experiment/{experiment_name}/convert",
            params={"variant": variant},
            headers=self._auth_headers(),
            timeout=5,
        )

    # ------------------------------------------------------------------
    # Rate limiting
    # ------------------------------------------------------------------

    def check_rate_limit(self, route: str, user_id: str = "", tenant_id: str = "") -> bool:
        """Returns True if the request is allowed, False if throttled. Fails open."""
        try:
            resp = requests.post(
                f"{self.base_url}/ratelimit/check",
                json={"route": route, "userId": user_id, "tenantId": tenant_id},
                timeout=2,
            )
            return resp.json().get("allowed", True)
        except Exception:
            return True  # fail open

    # ------------------------------------------------------------------
    # Circuit breaking
    # ------------------------------------------------------------------

    def get_circuit_breaker_state(self, route: str) -> str:
        """
        Returns the circuit breaker state for the given route.
        One of: "closed", "open", "half-open". Fails open (returns "closed")
        on any network or decode error.
        """
        try:
            resp = requests.get(
                f"{self.base_url}/circuitbreaker",
                params={"route": route},
                timeout=2,
            )
            return resp.json().get("state", "closed")
        except Exception:
            return "closed"  # fail open

    def check_circuit_breaker(self, route: str) -> tuple[bool, str]:
        """Returns (allowed, state) for the route. Fails open on errors."""
        try:
            resp = requests.post(
                f"{self.base_url}/circuitbreaker/check",
                json={"route": route},
                timeout=2,
            )
            data = resp.json()
            return data.get("allowed", True), data.get("state", "closed")
        except Exception:
            return True, "closed"

    def report_circuit_breaker(self, route: str, success: bool, latency_ms: int = 0) -> None:
        """
        Report a completed request outcome and latency to the control-plane
        so it can update the circuit breaker state for the given route.
        """
        try:
            requests.post(
                f"{self.base_url}/circuitbreaker/report",
                json={"route": route, "success": success, "latencyMs": latency_ms},
                headers=self._auth_headers(),
                timeout=2,
            )
        except Exception:
            pass  # non-critical reporting; swallow errors

File: sdk/featureflags_sdk.js
// Minimal Node.js SDK for Feature Flags Control Plane
// Requires: npm install node-fetch eventsource

const fetch = require('node-fetch');
const EventSource = require('eventsource');

class FeatureFlagsClient {
  constructor(baseUrl, env = 'dev', authToken = '') {
    this.baseUrl = baseUrl;
    this.env = env;
    this.authToken = authToken;
    this.flags = [];
  }

  authHeaders() {
    return this.authToken ? { Authorization: `Bearer ${this.authToken}` } : {};
  }

  // -------------------------------------------------------------------------
  // Flag listing & hot-reload
  // -------------------------------------------------------------------------

  async getAllFlags() {
    const res = await fetch(`${this.baseUrl}/flags/all?env=${this.env}`);
    if (res.ok) {
      this.flags = await res.json();
      return this.flags;
    }
    return [];
  }

  pollFlags(interval = 10000) {
    setInterval(() => this.getAllFlags(), interval);
  }

  startSSE(onUpdate) {
    const url = `${this.baseUrl}/flags/stream?env=${this.env}`;
    const es = new EventSource(url);
    const applyUpdate = async (event) => {
      try {
        this.flags = JSON.parse(event.data);
      } catch (_) {
        this.flags = await this.getAllFlags();
      }
      if (onUpdate) onUpdate(this.flags);
    };

    es.onmessage = (event) => {
      if (event.type === 'update') {
        void applyUpdate(event);
      }
    };
    es.addEventListener('update', (event) => {
      void applyUpdate(event);
    });
    es.addEventListener('ping', () => {});
    return es;
  }

  // -------------------------------------------------------------------------
  // Feature flag evaluation
  // -------------------------------------------------------------------------

  /**
   * Evaluate a feature flag for a user/tenant context.
   * Returns a boolean; fails open (returns false) on network errors.
   */
  async isEnabled(flagName, userId = '', tenantId = '', headers = {}) {
    try {
      const res = await fetch(`${this.baseUrl}/flags/${flagName}/evaluate`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ UserID: userId, TenantID: tenantId, Headers: headers }),
      });
      if (!res.ok) return false;
      const data = await res.json();
      return Boolean(data.enabled);
    } catch (_) {
      return false;
    }
  }

  // -------------------------------------------------------------------------
  // A/B experiments
  // -------------------------------------------------------------------------

  /**
   * Get the deterministic variant assigned to a user for an experiment.
   * Returns variant string, or empty string on error.
   */
  async getVariant(experimentName, userId) {
    try {
      const res = await fetch(
        `${this.baseUrl}/experiment/${experimentName}/variant?userId=${encodeURIComponent(userId)}`
      );
      if (!res.ok) return '';
      const data = await res.json();
      return data.variant || '';
    } catch (_) {
      return '';
    }
  }

  /**
   * Record a conversion event for an experiment variant.
   */
  async recordConversion(experimentName, variant) {
    try {
      await fetch(
        `${this.baseUrl}/experiment/${experimentName}/convert?variant=${encodeURIComponent(variant)}`,
        { method: 'POST', headers: this.authHeaders() }
      );
    } catch (_) {
      // non-critical; swallow
    }
  }

  // -------------------------------------------------------------------------
  // Rate limiting
  // -------------------------------------------------------------------------

  /**
   * Ask the control-plane whether the request is within rate limits.
   * Returns true (allowed) on network errors (fail open).
   */
  async checkRateLimit(route, userId = '', tenantId = '') {
    try {
      const res = await fetch(`${this.baseUrl}/ratelimit/check`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ route, userId, tenantId }),
      });
      if (!res.ok) return true;
      const data = await res.json();
      return Boolean(data.allowed);
    } catch (_) {
      return true; // fail open
    }
  }

  // -------------------------------------------------------------------------
  // Circuit breaking
  // -------------------------------------------------------------------------

  /**
   * Get the circuit breaker state for a route.
   * Returns one of: "closed", "open", "half-open".
   * Fails open (returns "closed") on errors.
   */
  async getCircuitBreakerState(route) {
    try {
      const res = await fetch(
        `${this.baseUrl}/circuitbreaker?route=${encodeURIComponent(route)}`
      );
      if (!res.ok) return 'closed';
      const data = await res.json();
      return data.state || 'closed';
    } catch (_) {
      return 'closed'; // fail open
    }
  }

  async checkCircuitBreaker(route) {
    try {
      const res = await fetch(`${this.baseUrl}/circuitbreaker/check`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ route }),
      });
      if (!res.ok) return { allowed: true, state: 'closed' };
      const data = await res.json();
      return { allowed: Boolean(data.allowed), state: data.state || 'closed' };
    } catch (_) {
      return { allowed: true, state: 'closed' };
    }
  }

  /**
   * Report a request outcome and latency to the control-plane so it can
   * update the circuit breaker state for the route.
   */
  async reportCircuitBreakerResult(route, success, latencyMs = 0) {
    try {
      await fetch(`${this.baseUrl}/circuitbreaker/report`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', ...this.authHeaders() },
        body: JSON.stringify({ route, success, latencyMs }),
      });
    } catch (_) {
      // non-critical; swallow
    }
  }
}

// Usage Example:
// const client = new FeatureFlagsClient('http://localhost:8080', 'dev');
// client.getAllFlags().then(console.log);
// client.startSSE(flags => console.log('Flags updated:', flags));
// client.isEnabled('dark-mode', 'alice').then(enabled => console.log('enabled:', enabled));
// client.getVariant('button-color', 'alice').then(v => console.log('variant:', v));
// client.checkRateLimit('/api/search', 'alice').then(ok => console.log('allowed:', ok));
// client.getCircuitBreakerState('/api/search').then(s => console.log('cb state:', s));

module.exports = FeatureFlagsClient;
